### PR TITLE
Add ecommerce tracking

### PR DESCRIPTION
**Pull request by JamesCGDS**
## What

This PR adds enhanced ecommerce tracking scripts that will gather data on pages such as search and push this data to the `dataLayer` where it can be accessed by Performance Analysts.

Caveat: on lines `90` and `91`, `internalLinksDomain` and `internalLinksDomainWithoutWww` are set directly because otherwise `isExternalLink` will return an error. This implementation is temporary and will be fixed following a refactor of the `ga4-link-tracker` code by @AshGDS (https://github.com/alphagov/govuk_publishing_components/pull/2961).

## Why

These changes are being made as part of our migration to GA4.

## Visual Changes

N/A

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,10 @@
   useful summary for people upgrading their application, not a replication
   of the commit log.
 
+## 30.5.3
```
Comment (AshGDS) [inline CHANGELOG.md:10]: This needs to be moved back to `## Unreleased` (I think a rebase causes this to happen as your change stays on the same line, but the heading changes)

File: app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+      var schema = this.populateEcommerceSchema(true, searchResult)
+
+      this.clearPreviousEcommerceObject()
+      window.dataLayer.push(schema)
+    }
+  }
+
+  Ga4EnhancedEcommerceTracker.prototype.populateEcommerceSchema = function (searchResultClicked = false, searchResult = null) {
+    // Limiting to 100 characters to avoid noise from extra longs search queries
+    // and to stop the size of the payload going over 8k limit.
```
Comment (AshGDS) [inline app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:68]: Is this is a comment from UA, if so I think it should be removed (the 8k limit part)

File: app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    var listTitle = this.module.getAttribute('data-list-title') || this.DEFAULT_LIST_TITLE
+    var startPosition = parseInt(this.module.getAttribute('data-ecommerce-start-index'), 10)
+    var ecommerceObject = {}
+
+    ecommerceObject.event = 'search_results'
+    ecommerceObject.search_results = {}
+    ecommerceObject.search_results.event_name = searchResultClicked && searchResult ? 'select_item' : 'view_item_list'
+    ecommerceObject.search_results.term = searchQuery
+    ecommerceObject.search_results.sort = variant
+    ecommerceObject.search_results.results = parseInt(this.getMetaTagContent('govuk:search-result-count'))
```
Comment (AshGDS) [inline app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:81]: Is it OK that if the meta tag doesn't exist, `parseInt(null)` would return `NaN` and push that to the dataLayer? Should it fallback to `ecommerceRows.length` here if the meta tag doesn't exist?

File: spec/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.spec.js [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+    searchResultsParentEl = document.createElement('div')
+    searchResultsParentEl.setAttribute('data-search-query', 'test-search-query')
+    searchResultsParentEl.setAttribute('data-module', 'ga4-enhanced-ecommerce-tracker')
+    searchResultsParentEl.setAttribute('data-ecommerce-variant', 'test-ecommerce-variant')
+    searchResultsParentEl.setAttribute('data-list-title', 'test-list-title')
+    searchResultsParentEl.setAttribute('data-ecommerce-start-index', '1')
+
+    searchResults = document.createElement('div')
+    searchResults.innerHTML =
+      '<a data-ecommerce-path="/coronavirus" data-ecommerce-row="1" data-ecommerce-index="1" data-track-category="navFinderLinkClicked" data-track-action="Search.1" data-track-label="/coronavirus" data-track-options="{&quot;dimension28&quot;:20,&quot;dimension29&quot;:&quot;Coronavirus (COVID-19): guidance and support&quot;}" class="gem-c-document-list__item-title   govuk-link" href="/coronavirus">Coronavirus (COVID-19): guidance and support</a>' +
```
Comment (AshGDS) [inline spec/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.spec.js:31]: Might be worth removing the data values and the css class which aren't needed for these elements, so it's easier to read

**Review by AshGDS — COMMENTED** [review]
Thanks @JamesCGDS - I've added some comments :+1:

File: app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+window.GOVUK.Modules = window.GOVUK.Modules || {};
+
+(function (Modules) {
+  'use strict'
+
+  function Ga4EnhancedEcommerceTracker (module) {
+    this.module = module
+  }
+
+  Ga4EnhancedEcommerceTracker.prototype.init = function (isNewPage = true) {
```
Comment (andysellick) [inline app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:11]: We don't normally pass parameters to the `init` function of a module like this, normally because Modules are called and initialised by the start Modules script. Instead, normal practice is to have another data attribute on the element in question and read it when the constructor is called, or in the init function. For example:

```HTML
<div data-module="Ga4EnhancedEcommerceTracker" data-is-new-page="false"></div>
```

```JavaScript
  function Ga4EnhancedEcommerceTracker (module) {
    this.module = module
    this.isNewPage = this.module.getAttribute('data-is-new-page') || true
  }
```

File: app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+      this.startModule = this.startModule.bind(this)
+      window.addEventListener('cookie-consent', this.startModule)
+    }
+  }
+
+  Ga4EnhancedEcommerceTracker.prototype.startModule = function () {
+    this.PIIRemover = new window.GOVUK.analyticsGA4.PIIRemover()
+    this.DEFAULT_LIST_TITLE = 'Site search results'
+
+    window.dataLayer = window.dataLayer || []
```
Comment (andysellick) [inline app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:27]: I think we should try to avoid initialising the dataLayer inside our code. Instead, I think we should be checking to see whether it exists as an early step, and backing out if it doesn't (on the assumption that GTM has failed to initialise for some reason, perhaps because cookie consent hasn't been given).

File: app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    if (this.isNewPage) {
+      this.module.addEventListener('click', this.handleClick.bind(this))
+    } else {
+      if (window.GOVUK.analyticsGA4.analyticsModules.PageViewTracker) {
+        var pageViewTracker = window.GOVUK.analyticsGA4.analyticsModules.PageViewTracker
+        window.dataLayer.push({
+          event: 'page_view',
+          page_view: {
+            location: pageViewTracker.getLocation(),
+            referrer: pageViewTracker.getReferrer()
```
Comment (andysellick) [inline app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:38]: I'm a bit confused by this section of code - it looks like you're getting the page view code and using part of it to create a new page view in here? I feel like this could increase maintenance if we decide to change the data structure of page views. Instead, could you call the page view code directly?

```JavaScript
var pageViewTracker = window.GOVUK.analyticsGA4.analyticsModules.PageViewTracker
if (pageViewTracker) {
  pageViewTracker.init()
}
```

File: app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    var ecommerceObject = {}
+
+    ecommerceObject.event = 'search_results'
+    ecommerceObject.search_results = {}
+    ecommerceObject.search_results.event_name = searchResultClicked && searchResult ? 'select_item' : 'view_item_list'
+    ecommerceObject.search_results.term = searchQuery
+    ecommerceObject.search_results.sort = variant
+    ecommerceObject.search_results.results = parseInt(this.getMetaTagContent('govuk:search-result-count'))
+    ecommerceObject.search_results.ecommerce = {}
+    ecommerceObject.search_results.ecommerce.items = []
```
Comment (andysellick) [inline app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:83]: Could this block be written out as an object instead of individual declarations? If it's possible I think it would help to make the structure of the object more plain.

```JavaScript
var ecommerceObject = {
  event: 'search_results',
  search_results: {
    event_name: searchResultClicked && searchResult ? 'select_item' : 'view_item_list',
    // etc.
  }
}
```

File: app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+    if (searchResultClicked && searchResult) {
+      ecommerceObject.search_results.ecommerce.items.push({
+        item_id: searchResult.getAttribute('data-ecommerce-path'),
+        item_name: searchResult.textContent,
+        item_list_name: listTitle,
+        index: this.getIndex(searchResult, Array.prototype.indexOf.call(ecommerceRows, searchResult)) + startPosition
```
Comment (andysellick) [inline app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:90]: Can you explain what this line does?

File: app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+  Ga4EnhancedEcommerceTracker.prototype.getMetaTagContent = function (metaTagName) {
+    var element = document.querySelector(`meta[name='${metaTagName}']`)
+    if (!element) {
+      return null;    
+    }
+    return element.getAttribute('content');
+  }
+
+  Modules.Ga4EnhancedEcommerceTracker = Ga4EnhancedEcommerceTracker
```
Comment (andysellick) [inline app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:132]: Since this gets referred to in a few places, would it be cleaner to call this something shorter? Maybe just `Ga4EcommerceTracker`?

File: spec/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.spec.js [hunk 11]
```diff
@@ -0,0 +1,295 @@
+/* eslint-env jasmine */
+
+describe('Google Analytics ecommerce tracking', function () {
```
Comment (andysellick) [inline spec/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.spec.js:3]: I'd consider merging the commit that adds this test with the commit that creates the code it refers to, since they're linked.

File: app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js [hunk 12]
```diff
@@ -40,7 +40,7 @@ window.GOVUK.Modules = window.GOVUK.Modules || {};
         })
       }
     }
-    
+
```
Comment (andysellick) [inline app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:43]: Could these linting fixes be merged with the commit that added these files?

**Review by andysellick — COMMENTED** [review]
Looks generally good, although it's a little hard to follow in places and I'm not sure it should be structured as one of our `GOVUK.Modules`. Let's have a chat and see what solution would be best.

File: CHANGELOG.md (hunk 1 above)
Comment (JamesCGDS) [reply CHANGELOG.md]: Done.

File: app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js (hunk 2 above)
Comment (JamesCGDS) [reply app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js]: I think the comment should stay because I think it does relate to GA. Good spot re the typo, I've fixed that now.

File: app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js (hunk 3 above)
Comment (JamesCGDS) [reply app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js]: We can't use `ecommerceRows.length` because that doesn't return the total result count. As per our discussion, I've changed this function to get the total results from the `h2`.

File: spec/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.spec.js (hunk 4 above)
Comment (JamesCGDS) [reply spec/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.spec.js]: Agreed, I've removed those.

File: app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+          }
+        })
+      }
+    }
+
+    this.trackSearchResults()
+  }
+
+  Ga4EnhancedEcommerceTracker.prototype.trackSearchResults = function () {
+    var schema = this.populateEcommerceSchema(this.module)
```
[the remark this answers is not in the document]
Comment (JamesCGDS) [reply app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js]: Good spot!

File: app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js (hunk 6 above)
Comment (JamesCGDS) [reply app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js]: Agreed, I've changed this to just be `if(window.dataLayer){ ... }`

File: app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js (hunk 7 above)
Comment (JamesCGDS) [reply app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js]: As per our discussion earlier, I've implemented your above suggestion with a view to refactor the `PageViewTracker` code at a later date to accommodate scenarios where we only need to push a few select values of the `pageView` object to the `dataLayer` rather than the whole thing.

File: app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js (hunk 8 above)
Comment (JamesCGDS) [reply app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js]: Agreed, this is much easier to read.

File: app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js (hunk 9 above)
Comment (JamesCGDS) [reply app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js]: As per our discussion earlier, this code has been refactored to be more readable following a suggestion from @AshGDS.

File: app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js (hunk 10 above)
Comment (JamesCGDS) [reply app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js]: Agreed. As per our discussion earlier, this will be addressed as part of a wider task to look at naming conventions.

File: spec/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.spec.js (hunk 11 above)
Comment (JamesCGDS) [reply spec/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.spec.js:3]: Done.

File: app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js (hunk 12 above)
Comment (JamesCGDS) [reply app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:20]: Done.

**Comment by JamesCGDS** [comment]
Thank you @andysellick and @AshGDS for your comments and input, it's much appreciated 👍  I've addressed those comments now, if you have any further feedback just give me a shout!

**Comment by AshGDS** [comment]
Looks good to me but I will leave it for Andy to approve :+1:

File: app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+      if (searchResultClicked && searchResult) {
+        ecommerceObject.search_results.ecommerce.items.push({
+          item_id: searchResult.getAttribute('data-ecommerce-path'),
+          item_name: searchResult.textContent,
+          item_list_name: listTitle,
+          index: this.getIndex(searchResult, startPosition)
+        })
+
+        GOVUK.analyticsGA4.analyticsModules.Ga4LinkTracker.internalLinksDomain = 'www.gov.uk/'
+        GOVUK.analyticsGA4.analyticsModules.Ga4LinkTracker.internalLinksDomainWithoutWww = 'gov.uk/'
```
Comment (andysellick) [inline app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:93]: Hopefully a quick one. Can you recap why these need to be set? And will it have any detrimental effect on the rest of the code once these two lines have run?

File: spec/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.spec.js [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+    searchResultsParentEl.setAttribute('data-module', 'ga4-enhanced-ecommerce-tracker')
+    searchResultsParentEl.setAttribute('data-ecommerce-variant', 'test-ecommerce-variant')
+    searchResultsParentEl.setAttribute('data-list-title', 'test-list-title')
+    searchResultsParentEl.setAttribute('data-ecommerce-start-index', '1')
+
+    searchResults = document.createElement('div')
+    searchResults.innerHTML =
+      '<a data-ecommerce-path="/coronavirus" data-ecommerce-row="1" data-ecommerce-index="1" href="/coronavirus">Coronavirus (COVID-19): guidance and support</a>' +
+
+      '<a data-ecommerce-path="/guidance/travel-abroad-from-england-during-coronavirus-covid-19" data-ecommerce-row="1" data-ecommerce-index="2" href="/guidance/travel-abroad-from-england-during-coronavirus-covid-19">Travel abroad from England during coronavirus (COVID-19)</a>' +
```
Comment (andysellick) [inline spec/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.spec.js:34]: Non blocking comment - it's good to base test data on real data but we can make our test data more readable by shortening things to the minimum i.e. the text and href of these links could be much shorter - as long as they're unique.

**Review by andysellick — APPROVED** [review]
Looks great, well done!

File: spec/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.spec.js (hunk 15 above)
Comment (JamesCGDS) [reply spec/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.spec.js]: Cool, I've just pushed up a change that will shorten those test links.

File: app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js (hunk 14 above)
Comment (JamesCGDS) [reply app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js]: I’ve just pushed up a change that removes these. Initially, we were having problems without setting it but I’ve just tested this locally and everything works fine without them. The only place where we’d need to keep them is in the tests as the link tracker isn’t initialised there.
